### DISCUSSION TITLE
Limits of open_mfdataset for many netcdf files

### DISCUSSION THREAD

**Discussion by betolink** [pydata/xarray#8925] [Q&A]
I'm curious to know, what's the max number of NetCDF/HDF5 files (remote or local) that someone has opened with xarray's `xr.open_mfdataset(files, parallel=True)`?  This is using a Dask cluster.  Context: I'm opening ~4000 small files and I'm seeing a lot of memory usage just on the opening step, these are small files (3MB each), even loaded into memory they should fit on my instance.

**Comment by douglatornell** [comment]
We routinely process ~30 netCDF4 files in the 25M-75M size range to calculate month-averaged fields from day-averaged fields. Less frequently, we process 365 netCDF4 files in the 200M-2G size range to construct year-long datasets from daily hour-averaged fields. The files are all locally stored on a Linux file system. We use a `processes=True` dask cluster. `parallel=True` does not work well for us. That and more than 1 thread per process seemed to be the cause of random netCDF and HDF5 errors as described in https://github.com/xCDAT/xcdat/issues/561#issuecomment-1969470260

**Reply by ZZMitch** [reply]
Just wanted to note that I was having issues with `open_mfdataset` in my workflow, but just setting `parallel=False` seemed to fix it. 

To be more specific, I had ~160 locally saved NetCDF files (spatial tiles) that I wanted to merge into a set of wide_area tifs (one for each variable in NetCDF file). With older versions of `xarray` last year this was working fine with  `open_mfdataset` and `parallel=True` but this year my TIF outputs were missing some tiles and some tiles were in the wrong location. I switched `parallel=False`, ran the same code and got the correct output. It also did not seem to impact processing speed much - so I am happy. I am using `dask LocalCluster` as well. 

I will keep the rest of this thread in mind though, since I do this type of wide-area merging a lot.

**Reply by ZZMitch** [reply]
For reference, with `parallel=True`:
![image](https://github.com/user-attachments/assets/efc7d98f-ad81-42ab-9d1f-ef3f5862d134)
Note some tiles missing and in wrong locations. 

And after setting to `parallel=False`:
![image](https://github.com/user-attachments/assets/eddbd8a6-f52a-4689-a1ff-7db55336855e)
All looks good!

**Reply by dcherian** [reply]
This is concerning. Are you able ot track it down?

**Comment by dcherian** [comment]
I've done O(10,000) files with a big-ish distributed cluster.

Your memory usage might result from useless comparisons. See the "Note" in https://docs.xarray.dev/en/stable/user-guide/io.html#reading-multi-file-datasets (xref https://github.com/pydata/xarray/issues/8778)

**Comment by betolink** [comment]
Thanks @dcherian, this is somewhat related to a comment you made not long ago on how to once that we have these many files open, operations should be faster if we use flox. I think your suggestion of changing the default values in `open_mfdataset` makes total sense to me. 

For context, I'm prototyping a way in which earthaccess will take advantage of knowing how compliant/regular a given dataset is and use this information to speed up access by inferring sensible defaults in xarray and using the right caching strategy (via fsspec).

**Reply by dcherian** [reply]
Very nice, we'd be happy to publish a blog post on this topic at the Xarray blog if you want.

**Comment by ghiggi** [comment]
Hi @betolink. I often used `xr.open_mfdataset(files, parallel=True)` to open O(100,000) files.

In my experience was very important to:
- increase the maximum allowed number of opened file with `ulimit -n <number>`. The default on Linux is often `1024`
- set `os.environ["HDF5_USE_FILE_LOCKING"] = "FALSE"` before the imports (or as environment variable) 
- Wtih some servers/NFS configurations, to set `dask.config.set({'distributed.worker.use-file-locking': 'False'})` 
- constrain the dask cluster to `threads_per_worker=1` to avoid netcdf locking

If the reason to open lot of files to then rechunk it (i.e. with rechunker) you should then also pay attention to the max chunk size you are gonna to create and the compressor you use. Some compressor have an upper limit on the chunk size they can deal with ... and cause some cryptic error messages or failure of the rechunking.

I saw the recent efforts you put on `earthaccess` and I tackled similar problems to facilitate the access and analysis of the GPM data archive nested group HDFs with xarray (see [https://gpm-api.readthedocs.io/en/latest/](https://gpm-api.readthedocs.io/en/latest/)). We can have a chat in the coming days if you want ;)

**Reply by dcherian** [reply]
Great answer. Thanks for taking the time to write this up!

**Reply by betolink** [reply]
I'd love to compare notes @ghiggi! I'll send you a message on Twitter, I have some time this Friday or next week.

**Reply by dcherian** [reply]
Can I suggest the Pangeo weekly meeting as a venue for this discussion please? I bet many are interested, and public discussions help reduce bus factor, and build up a broad pool of expertise.

**Reply by betolink** [reply]
That's a great idea @dcherian!

**Reply by ghiggi** [reply]
Hi @betolink and @dcherian. I will try to attend this evening meeting ;)
